Support for Actor timers, and fix bug in FSM, #15733

patriknw opened the pull request:
* backport of the timers from Akka Typed, #16742
* also fixed a small bug in FSM timers, which could result in that
  a timer from a previous incarnation was let through to new
  incarnation after restart
* no more need for the complicated "how to" section in docs of
  how to schedule periodic messages

Refs #15733

patriknw commented on project/MiMa.scala:
```diff
@@ -1159,6 +1160,14 @@ object MiMa extends AutoPlugin {
+//        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.persistence.fsm.PersistentFSM#Timer.apply"),
+//        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.persistence.fsm.PersistentFSM#Timer.copy"),
+//        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.persistence.fsm.PersistentFSM#Timer.this")
+//      )    
```
@2m I can't get this to work. I have thrown those lines at several places but nothing sticks (I don't understand).

patriknw commented on akka-actor-tests/src/test/scala/akka/actor/TimerSpec.scala:
```diff
@@ -0,0 +1,269 @@
+      probe.expectMsg(GotPostStop(false))
+    }
+
+    "discard timers from old incarnation after restart, alt 1" taggedAs TimingTest in {
```
this is the reproducer for the FSM bug

patriknw commented on akka-actor/src/main/scala/akka/actor/Actor.scala:
```diff
@@ -272,6 +272,7 @@ class ActorInterruptedException private[akka] (cause: Throwable) extends AkkaExc
  */
 @SerialVersionUID(1L)
 final case class UnhandledMessage(@BeanProperty message: Any, @BeanProperty sender: ActorRef, @BeanProperty recipient: ActorRef)
+  extends NoSerializationVerificationNeeded
```
I noticed that if a message that is not serializable is published as unhandled then the unhandled message is reported as not serializable and hiding the root cause

2m commented:
Added a commit that shuffles filters around and makes MiMa happy now. There were two things that needed to be fixed:

* fitlers for `akka.actor.FSM#Timer.*` were missing
* after adding 2.4.18 to the list, some of the BC issues reappeared because filters for those were in the 2.4.17 filter list. These were moved to `bcIssuesBetween24and25`.

patriknw commented on akka-actor/src/main/scala/akka/actor/Timers.scala:
```diff
@@ -0,0 +1,134 @@
+   * `TimerScheduler` is not thread-safe, i.e. it must only be used within
+   * the actor that owns it.
+   */
+  def createTimerScheduler(context: AbstractActor.ActorContext): TimerScheduler =
```
remove this, it's not useful without the intercept of TimerMessage

patriknw commented:
I should probably take the mima change here into a separate PR so that we can get mima working in master.

patriknw committed:
Support for Actor timers, and fix bug in FSM, #15733

* backport of the timers from Akka Typed, #16742
* also fixed a small bug in FSM timers, which could result in that
  a timer from a previous incarnation was let through to new
  incarnation after restart
* no more need for the complicated "how to" section in docs of
  how to schedule periodic messages

patriknw commented:
I have rebased this and it's ready, please review.

ktoso commented on akka-docs/src/main/paradox/scala/scheduler.md:
```diff
@@ -10,6 +10,10 @@ You can schedule sending of messages to actors and execution of tasks
 `cancel` on to cancel the execution of the scheduled operation.
 
+When scheduling periodic or single messages in an actor to itself it is recommended to
+use the @ref:[Actor Timers](actors.md#actors-timers) instead of using the ``Scheduler``
```
hehe good ol restructured text style snuck in here, single ` is fine

ktoso reviewed (approved):
LGTM in general :) Happy to see this included :)

patriknw commented on akka-actor/src/main/scala/akka/actor/Timers.scala:
```diff
@@ -0,0 +1,134 @@
+  /**
+   * Check if a timer with a given `key` is active.
+   */
+  def isTimerActive(key: Any): Boolean
```
what would be the meaning of such type? would it be the scheduled message type? I think it's just an identifier. We could consider if it makes sense for Typed, but for untyped I think it just makes it complicated.

ktoso replied:
I was thinking to have the `key` typed, to not accidentally use some wrong key when getting the timer... Not sure if very worth it

ktoso replied:
Just that the Any feels a bit weird, ok to keep as is, but something to think about hm

patriknw commented on akka-actor/src/main/scala/akka/actor/dungeon/TimerSchedulerImpl.scala:
```diff
@@ -0,0 +1,107 @@
+
+  private val log = Logging(ctx.system, classOf[TimerScheduler])
+  private var timers: Map[Any, Timer] = Map.empty
+  private val timerGen = Iterator from 1
```
yes

patriknw commented:
Incorporated feedback from @ktoso 
One more reviewer, please. @raboof ?

raboof commented on akka-actor/src/main/scala/akka/actor/Timers.scala:
```diff
@@ -0,0 +1,119 @@
+   * previous timer is not received, even though it might already be enqueued
+   * in the mailbox when the new timer is started.
+   */
+  def startPeriodicTimer(key: Any, msg: Any, interval: FiniteDuration): Unit
```
Should we have both a `timeout`/`delay` and an `interval` like on `scheduler`?

patriknw replied:
that is very rarely used, as far as I know, and with this it's easy to first schedule a single timer and then the period after that, if different initial delay is needed

raboof commented on akka-actor-tests/src/test/scala/akka/actor/TimerSpec.scala:
```diff
@@ -0,0 +1,269 @@
+}
+
+abstract class AbstractTimerSpec extends AkkaSpec("""
+  akka.loglevel = DEBUG
```
Is this (still) needed?

patriknw replied:
no, I'll remove

raboof commented on akka-actor/src/main/scala/akka/actor/dungeon/TimerSchedulerImpl.scala:
```diff
@@ -0,0 +1,111 @@
+    val timerMsg = TimerMsg(key, nextGen, this)
+    val task =
+      if (repeat)
+        ctx.system.scheduler.schedule(timeout, timeout, ctx.self, timerMsg)(ctx.dispatcher)
```
Would be nice to also be able to schedule a recurring message without delay

patriknw replied:
without delay, you mean sending the first message immediately?

raboof commented on akka-docs/src/main/paradox/scala/actors.md:
```diff
@@ -905,6 +905,29 @@ Messages marked with `NotInfluenceReceiveTimeout` will not reset the timer. This
 `ReceiveTimeout` should be fired by external inactivity but not influenced by internal activity,
 e.g. scheduled tick messages.
 
+<a id="actors-timers"></a>
```
Indeed there should be automatic anchors for all headings (though we don't currently check for dangling references, https://github.com/lightbend/paradox/issues/120

raboof commented on akka-docs/src/main/paradox/scala/howto.md:
```diff
@@ -122,47 +122,7 @@ The pattern is described [Discovering Message Flows in Actor System with the Spi
-:  @@snip [SchedulerPatternTest.java]($code$/java/jdocs/pattern/SchedulerPatternTest.java) { #schedule-receive }
-
-@@@ div { .group-java }
+See @ref:[Actor Timers](actors.md#actors-timers)
```
Perhaps remove the section here entirely?

patriknw replied:
I kept it because I think we have linked to it rather much.

raboof commented on akka-docs/src/main/paradox/scala/scheduler.md:
```diff
@@ -10,6 +10,10 @@ You can schedule sending of messages to actors and execution of tasks
 
+When scheduling periodic or single messages in an actor to itself it is recommended to
+use the @ref:[Actor Timers](actors.md#actors-timers) instead of using the `Scheduler`
+directly.
```
I like how you don't have to go to `context.system.scheduler` for this common case anymore :+1:

raboof commented on akka-docs/src/main/paradox/scala/actors.md:
```diff
@@ -905,6 +905,29 @@ Messages marked with `NotInfluenceReceiveTimeout` will not reset the timer. This
 
+<a id="actors-timers"></a>
+
+## Timers, scheduled messages
```
actors.md is getting quite long, perhaps we should move some things to their own file... not sure what though :).

patriknw replied:
yes, lot's of that restructuring would be nice, but that is for another time

raboof reviewed (commented):
Generally LGTM, but consider if we also want a `delay` parameter on `startPeriodicTimer`. Apart from that good change, reasonable test, neat docs!

patriknw commented:
rebased and removed DEBUG, will merge after another pr validation

dwijnand commented on akka-actor/src/main/scala/akka/actor/Timers.scala:
```diff
@@ -0,0 +1,134 @@
+  /**
+   * Check if a timer with a given `key` is active.
+   */
+  def isTimerActive(key: Any): Boolean
```
https://github.com/akka/akka/pull/25865

dwijnand commented on akka-actor/src/main/scala/akka/actor/dungeon/TimerSchedulerImpl.scala:
```diff
@@ -0,0 +1,107 @@
+      case None ⇒
+        // it was from canceled timer that was already enqueued in mailbox
+        log.debug("Received timer [{}] that has been removed, discarding", timerMsg.key)
+        OptionVal.None // message should be ignored
```
Need to forward-port this back-port, back into Akka Typed that's still using `null` like a wild animal.
